src/algorithms/fardetectors/PolynomialMatrixReconstruction.cc

```diff
@@ -0,0 +1,343 @@
+      numBeamProtons++;
+    }
+    if (p.getGeneratorStatus() == 4 &&
+        p.getPDG() == 2112) { //look for "beam" neutron (for deuterons)
+      runningMomentum += p.getMomentum().z;
+      numBeamProtons++;
+    }
+    if (p.getGeneratorStatus() == 1 && p.getPDG() == 2212) {
+      trace("Final-state proton MCParticle momentum --> px = {}, py = {}, pz = {}",
+            p.getMomentum().x, p.getMomentum().y, p.getMomentum().z);
+    }
+  }
+
+  if (numBeamProtons == 0) {
+    if (m_cfg.requireBeamProton) {
+      critical("No beam protons found");
+      throw std::runtime_error("No beam protons found");
+    }
+    return;
+  }
+
+  nomMomentum = runningMomentum / numBeamProtons;
+
+  trace("nomMomentum extracted from beam protons --> nomMomentum = {}", nomMomentum);
+
+  double nomMomentumError = 0.05;
+
+  //method for polynomial method of reconstruction
+  //
+  // 1) use the hit information to find the correct X_L value needed to evaluate the matrix
+  // 2) evaluate the correct matrix for x and y reconstruction
+  // 3) invert the matrices and extract the needed deltaP/p and theta_x,y values
+  // 4) use 3) to calculate the momentum vector
+  // 5) save output
+  //
+  // matrices are as follows
+  //
+  // | a0  a1 |   | theta_x_ip |   | x_rp       |
+  // |        | x |            | = |            |
+  // | b0  b1 |   | deltap/p   |   | theta_x_rp |
+  //
+  // | c0  c1 |   | y_ip       |   | y_rp       |
+  // |        | x |            | = |            |
+  // | d0  d1 |   | theta_y_ip |   | theta_y_rp |
+  //
+  // the "y_ip" part is a bit extraneous is more for some level of testing - we generally "assume" the vectors come from
+  // (x, y, z) = (0, 0, 0), because we don't really have a choice.
+  //
+  // We will test how vertex smearing contributes to overall worsening of the resolution with the generally "correct" matrix being used
+  // allowing us to isolate the various smearing effects.
+  //
+
+  bool valid_energy_found = false;
+  for (const PolynomialMatrixConfig& poly_matrix_configs : m_cfg.poly_matrix_configs) {
+    if (std::abs(poly_matrix_configs.nomMomentum - nomMomentum) / poly_matrix_configs.nomMomentum <
+        nomMomentumError) {
+      if (valid_energy_found) {
+        error("Conflicting matrix values matching momentum {}", nomMomentum);
+      }
+      valid_energy_found = true;
+      nomMomentum        = poly_matrix_configs.nomMomentum;
+    }
+  }
+  if (not valid_energy_found) {
+    if (m_cfg.requireValidBeamEnergy) {
+      critical("No tune beam energy found - cannot acquire lookup table");
+      throw std::runtime_error("No valid beam energy found, cannot reconstruct momentum");
+    }
+    return;
+  }
+
+  //xL table filled here from LUT -- Graph2D used for nice interpolation functionality and simple loading of LUT file
+
+  TGraph2D* xLGraph = new TGraph2D(
+      Form("calibrations/RP_60_xL_100_beamEnergy_%.0f.xL.lut", nomMomentum), "%lf %lf %lf");
```

Comment:
```suggestion
  std::unique_ptr<TGraph2D> xLGraph { new TGraph2D(
      Form("calibrations/RP_60_xL_100_beamEnergy_%.0f.xL.lut", nomMomentum), "%lf %lf %lf") };
```

Reply:
Ah, this is helpful - I actually didn't know how to do this. If the beam momentum changes within the event, we have bigger problems - it really should not ever happen --> this is a good solution for now.